### PR TITLE
added integration tests

### PR DESCRIPTION
**Pull request by andreeaflorescu**
Added integration tests that cover single & multi-threaded
applications.

The integration tests make use of common subscribers defined
in the tests/common module.

The tests are also linked as examples in the README and should
be a good starting point for using the event-manager.

Fixes: https://github.com/rust-vmm/event-manager/issues/5

Signed-off-by: Andreea Florescu <fandree@amazon.com>

### CODE REVIEW

File: tests/basic_event_manager.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        for i in subscriber_index {
+            let subscriber = self
+                .event_manager
+                .subscriber_mut(self.subscribers_id[*i])
+                .unwrap();
+            subscriber.trigger_event();
+        }
+    }
+
+    fn clear_events_for(&mut self, subscriber_index: &Vec<usize>) {
```
Comment (alexandruag) [inline tests/basic_event_manager.rs:51]: Just curious if there's any special reason why `subscriber_index` is a `Vec` here, and a slice in the previous function? Also, something like `subscriber_indices` might be a more appropriate name.

File: tests/basic_event_manager.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            let subscriber = self.event_manager.subscriber_mut(*subscriber_id).unwrap();
+            result.push(subscriber.counter());
+        }
+
+        result
+    }
+}
+
+impl Drop for App {
+    fn drop(&mut self) {
```
Comment (alexandruag) [inline tests/basic_event_manager.rs:73]: I think the subscribers within the event manager get dropped/removed automatically when the `App` object is dropped (because the `event_manager` member will be dropped, which drops the `EventManager::subscribers` collection and its members). Is there any reason for explicitly implementing `Drop` in this example?

File: tests/basic_event_manager.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    // Random subscribers, in the sense that I randomly picked those numbers :)
+    let triggered_subscribers: Vec<usize> = vec![1, 3, 50, 97];
+    app.inject_events_for(&triggered_subscribers);
+    app.run();
+
+    let counters = app.get_counters();
+    for i in 0..100 {
+        assert_eq!(counters[i], 1 & (triggered_subscribers.contains(&i) as u64));
```
Comment (alexandruag) [inline tests/basic_event_manager.rs:94]: This is subjective, but I think the `bool` to `u64` cast is a bit ugly and potentially confusing here, and maybe it's nicer to be more verbose, but clearer along the lines of:
```rust
if triggered_subscribers.contains(&i) {
    assert_eq!(counters[i], 1);
} else {
    assert_eq!(counters[i], 0);
}
```

File: tests/common.rs [hunk 4]
```diff
@@ -0,0 +1,273 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+// For some reason clippy & check think that the public methods in this file are not used
```
Comment (alexandruag) [inline tests/common.rs:4]: Code in the `tests` folder is treated differently than normal source files in some respects. The [documentation](https://doc.rust-lang.org/book/ch11-03-test-organization.html#the-tests-directory) give more details, and also talks about the case when there's a `common` submodule just like we're trying to do here. I haven't tried it out, but maybe it fixes this problem.

File: tests/common.rs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+impl Default for Counter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// A dummy subscriber that increments a counter whenever it processes
+// a new request.
+pub struct CounterSubscriber(pub Counter);
```
Comment (alexandruag) [inline tests/common.rs:83]: Just curious, why is the inner `Counter` `pub` if we also implement the `Deref` traits like that?

File: tests/common.rs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+            EventSet::ERROR => {
+                eprintln!("Got error on the monitored event.");
+            }
+            EventSet::HANG_UP => {
+                event_ops
+                    .remove(events)
+                    .unwrap_or(eprintln!("Encountered error during cleanup"));
+                panic!("Cannot continue execution. Associated fd was closed.");
+            }
+            _ => {}
```
Comment (alexandruag) [inline tests/common.rs:120]: Is this branch not expected to happen? Or do we actually want to ignore this possibility? Either way, for the sake of clarity and establishing best practices I guess we should add a comment describing why we don't do anything and/or add some sort of token message as if we logged something.

File: tests/common.rs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+impl CounterSubscriber {
+    pub fn new() -> Self {
+        CounterSubscriber(Counter::new())
+    }
+}
+
+impl MutEventSubscriber for CounterSubscriber {
+    fn process(&mut self, events: Events, event_ops: &mut EventOps) {
+        match events.event_set() {
```
Comment (alexandruag) [inline tests/common.rs:107]: In general, I think it's possible to receive multiple events at the same time (don't remember which combinations are valid and which aren't :( ). If that's the case, I think we should check for inclusion rather than equality when looking at the event_set maybe. Just raising this so we check what the best approach is. I guess this also applies to the other similar code snippets.

File: tests/multi_threaded.rs [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            // When the event manager loop exits, it will return the number of events it
+            // handled.
+            let ev_count = event_manager.run().unwrap();
+            // We are expecting the ev_count to be:
+            // 4 = 3 (events triggered from data_subscriber) + 1 (event from counter_subscriber).
+            assert_eq!(ev_count, 4);
+        }
+
+        // One really important thing is to always remove the subscribers once you don't need them
+        // any more.
```
Comment (alexandruag) [inline tests/multi_threaded.rs:42]: Maybe it's worth mentioning this is especially important when the `EventManger` keeps running, because if it also gets dropped, then subscribers are dropped as well anyway.

File: tests/multi_threaded.rs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    // this is the case in this example as well.
+    let mut event_manager = EventManager::<Arc<dyn EventSubscriber + Send + Sync>>::new()
+        .expect("Cannot create event manager");
+
+    let subscriber = Arc::new(CounterInnerMutSubscriber::new());
+    subscriber.trigger_event();
+
+    // Let's just clone the subscriber before adding it to the event manager. This will allow us
+    // to use it from this thread without the need to call into EventManager::subscriber_mut().
+    let subscr_id = event_manager.add_subscriber(subscriber.clone());
```
Comment (alexandruag) [inline tests/multi_threaded.rs:131]: Not really important, but `subscr_id` sounds jarring somehow :-s How about `sub_id`? :-s

Comment (andreeaflorescu) [reply tests/multi_threaded.rs:131]: I really avoided sub_id on purpose. `sub` for me means substitution, not subscriber. I'll just name it `subscriber_id`. I will assume humans have big enough screens nowadays to not be bother by a 13 chars long name :D

File: tests/multi_threaded.rs (hunk 8 above)
Comment (andreeaflorescu) [reply tests/multi_threaded.rs:42]: I'll make it more explicit.

File: tests/common.rs (hunk 7 above)
Comment (andreeaflorescu) [reply tests/common.rs:107]: Hmm, I'll double check that.

File: tests/common.rs (hunk 5 above)
Comment (andreeaflorescu) [reply tests/common.rs:83]: I was initially lazy. I'll drop the pub.

**Comment by andreeaflorescu** [comment]
> The `common` module should be in `common/mod.rs` instead. Otherwise, `cargo test` will treat any function in it as a test.

cool, didn't know that. i will update it and then maybe I'll also not get the weird warnings anymore :))

File: tests/common.rs (hunk 4 above)
Comment (andreeaflorescu) [reply tests/common.rs:4]: it doesn't look like it is fixing the problem :(

File: tests/basic_event_manager.rs (hunk 2 above)
Comment (andreeaflorescu) [reply tests/basic_event_manager.rs:73]: `EventManager` does not implement `Drop` (maybe it should), but this is not the point. What I am trying to exemplify here is that once you are done with the subscriber, you need to call remove. Remove is not just removing things from the vector collection (which is something that rust does anyway), but also calls `epoll_ctl_del` to let `epoll` know that it doesn't need to trigger events for those fds. Does it make sense?

File: tests/multi_threaded.rs (hunk 8 above)
Comment (andreeaflorescu) [reply tests/multi_threaded.rs:42]: Another thing to keep in mind is that remove still needs to be called even when the `EventManager` does not run anymore. Should we actually implement Drop on EventManger instead?

File: tests/basic_event_manager.rs (hunk 3 above)
Comment (andreeaflorescu) [reply tests/basic_event_manager.rs:94]: I prefer it as it is now.

File: tests/basic_event_manager.rs (hunk 2 above)
Comment (alexandruag) [reply tests/basic_event_manager.rs:73]: Sure makes sense; maybe some comments in the code can help drive this point further. The example itself is a bit contrived because all these things happen by default (i.e. without an explicit `Drop` implementation for `App`), but explanations add any potentially missing context. On a related note, regardless of whether a type has  a `Drop` implementation  or not, its members are recursively dropped when a value of that type goes out of scope anyway (some more info [here](https://doc.rust-lang.org/std/ops/trait.Drop.html)).

File: tests/multi_threaded.rs (hunk 8 above)
Comment (alexandruag) [reply tests/multi_threaded.rs:42]: It doesn't look like we need to have an explicit `Drop` implementation for `EventManager` (also please see the other comment related to `Drop`).

File: tests/basic_event_manager.rs (hunk 2 above)
Comment (aghecenco) [reply tests/basic_event_manager.rs:73]: Maybe move the `for` loop that removes subscribers in a non-`Drop` function (`App::stop_listening()` or something like that). Make it explicit that `remove_subscribers` is necessary (recommended?) when you don't care about those events anymore - maybe your app has moved on to a stage where events are irrelevant.

File: tests/common/mod.rs [hunk 10]
```diff
@@ -0,0 +1,278 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+// For some reason clippy & check think that the public methods in this file are not used
+// (even though they are). Let's allow dead code so that we don't get warnings all the time.
+// This is a bit sad, hopefully it gets fixed in more recent rust versions.
```
Comment (aghecenco) [inline tests/common/mod.rs:6]:
```suggestion
// This is a bit sad, hopefully it gets fixed in newer rust versions.
```

File: tests/common/mod.rs [hunk 11]
```diff
@@ -0,0 +1,278 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+// For some reason clippy & check think that the public methods in this file are not used
```
Comment (aghecenco) [inline tests/common/mod.rs:4]: What's the specific complaint? Why don't we see it in firecracker?

File: tests/common/mod.rs [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+///
+/// 2. CounterSubscriberWithData:
+///     - a dummy subscriber that increments a counter on events
+///     - this subscriber takes care of multiple events and makes use of `Events::with_data` so
+///       that in the `process` function it identifies the trigger of an event using the data
+///       instead of the file descriptor
+///     - it has to be explicitly mutated; for this reason it implements `MutEventSubscriber`.
+///
+/// 3. CounterInnerMutSubscriber:
+///     - a dummy subscriber that increments a counter on events
```
Comment (aghecenco) [inline tests/common/mod.rs:23]: nit: I think ``` - a dummy subscriber that increments a counter on events``` is superfluous as it applies to all subscribers; it brings more value to mention under each subscriber type what it does differently from the others

File: tests/multi_threaded.rs [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    let counter_subscriber = Arc::new(Mutex::new(CounterSubscriber::new()));
+    counter_subscriber.lock().unwrap().trigger_event();
+
+    // Saving the ids allows to modify the subscribers in the event manager context (i.e. removing
+    // the subscribers from the event manager loop).
+    let ds_id = event_manager.add_subscriber(data_subscriber);
+    let cs_id = event_manager.add_subscriber(counter_subscriber);
+
+    let thread_handle = thread::spawn(move || {
+        // We cannot run the event manager loop for ever. Let's just run it for 100 iterations.
```
Comment (aghecenco) [inline tests/multi_threaded.rs:31]: Yes we can but we choose not to :smile: I'd rephrase to say that 100 runs are *enough* because we don't want an infinite test

File: tests/multi_threaded.rs [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+// Showcase how you can manage a single type of subscriber using the same event manager
+// in a multithreaded context.
+#[test]
+fn test_one_type_subscriber() {
+    let mut event_manager =
+        EventManager::<CounterSubscriberWithData>::new().expect("Cannot create event manager");
+
+    // The `CounterSubscriberWithData` expects to receive a number as a parameter that represents
+    // the number it can use as its inner Events data.
```
Comment (aghecenco) [inline tests/multi_threaded.rs:57]:
```suggestion
    // The `CounterSubscriberWithData` expects to receive the number it can use as its inner `Events` data as a parameter.
```

File: tests/multi_threaded.rs [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        let data_subscriber_1 = event_manager.subscriber_mut(ds_id_1).unwrap();
+        data_subscriber_1.trigger_all_counters();
+
+        let data_subscriber_2 = event_manager.subscriber_mut(ds_id_2).unwrap();
+        data_subscriber_2.trigger_all_counters();
+    }
+
+    let thread_handle = thread::spawn(move || {
+        // We cannot run the event manager loop for ever. Let's just run it for 100 iterations.
```
Comment (aghecenco) [inline tests/multi_threaded.rs:81]: Same nit as earlier

File: tests/common/mod.rs (hunk 11 above)
Comment (andreeaflorescu) [reply tests/common/mod.rs:4]: I really have no idea what's the difference between firecracker and what I am doing here.
Without this I get the following warnings:
```
warning: method is never used: `new`
   --> tests/common/mod.rs:149:5
    |
149 |     pub fn new(first_data: u32) -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: method is never used: `trigger_all_counters`
   --> tests/common/mod.rs:161:5
    |
161 |     pub fn trigger_all_counters(&mut self) {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: method is never used: `get_all_counter_values`
   --> tests/common/mod.rs:167:5
    |
167 |     pub fn get_all_counter_values(&self) -> Vec<u64> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: method is never used: `new`
   --> tests/common/mod.rs:236:5
    |
236 |     pub fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^

warning: method is never used: `trigger_event`
   --> tests/common/mod.rs:243:5
    |
243 |     pub fn trigger_event(&self) {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: method is never used: `clear_event`
   --> tests/common/mod.rs:247:5
    |
247 |     pub fn clear_event(&self) {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: method is never used: `counter`
   --> tests/common/mod.rs:251:5
    |
251 |     pub fn counter(&self) -> u64 {
```

File: tests/common/mod.rs (hunk 12 above)
Comment (andreeaflorescu) [reply tests/common/mod.rs:23]: Hmmm, I think the differences are already documented. Do you think something more is needed? I wouldn't delete the first line with what the subscriber does because I am thinking that we might add more subscriber types in the future, and it would make sense to have a short description of them as a first comment.

File: tests/multi_threaded.rs (hunk 8 above)
Comment (andreeaflorescu) [reply tests/multi_threaded.rs:43]: The last comment is super confusing for me. Why wouldn't we need explicit drop for eventmanager? Who else calls remove_subscriber if not the event_manager who has ownership of the subscribers? `remove_subscriber` could also be called by the user of the library, but at the scope of this code, event manager should manage its own cleanup, right?

File: tests/basic_event_manager.rs (hunk 2 above)
Comment (andreeaflorescu) [reply tests/basic_event_manager.rs]: I moved the for loop to a different function, added more comments to what's the deal with it. This conversation is also a bit confusing to me, so I am not sure that's what you were asking for or not :))

Please take another look.
